### DISCUSSION TITLE
What does this error mean? (unregistered dependency)

### DISCUSSION THREAD

**Discussion by memxor** [aptos-labs/aptos-developer-discussions#53] [Questions] [answered]
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/059e6e2f-cf8f-458f-876b-89752447e2d7)

**Comment by 0x-j** [comment]
Do you have the source code of the dependency `aptos_horse_publisher_signer`  you declared? Whenever you declare a dependency in move.toml, you need to have the source code available. Checkout this [answer](https://github.com/aptos-labs/aptos-developer-discussions/discussions/36#discussioncomment-8710173) on how to get the source code for your dependency.

**Reply by 0x-j** [reply]
Ok, is [this](https://github.com/IntoTheVerse/APTHorse-contracts) the repo you have issue now? Let me try publishing it.

**Reply by memxor** [reply]
Yes

**Reply by 0x-j** [reply]
Are you publishing to the randomnet cuz randomness module is only available on randomnet? If you use `aptos init` to create the publisher account, you can init an account on randomnet like this 
```sh
aptos init --profile randomnet-1 --network custom --rest-url https://fullnode.random.aptoslabs.com --faucet-url https://faucet.random.aptoslabs.com
```
When I try to publish to randomnet i got a different error probably due to bug in contract
```sh
BUILDING Horse_Race_Aptos
package size 10343 bytes
{
  "Error": "Simulation failed with status: Move abort in 0x1::resource_account: 0x60001"
}
```

**Reply by memxor** [reply]
I've done this already `aptos init --profile randomnet-1 --network custom --rest-url https://fullnode.random.aptoslabs.com --faucet-url https://faucet.random.aptoslabs.com`.. 

I don't understand what you mean by declare dependency, cause all my files are in source, I'm not getting anything from outside. It's all in the same module.

**Comment by gregnazario** [comment, marked as answer]
This error means your module's address doesn't match the account that is publishing the code.
```bash
{
  "Error": "Simulation failed with status: Transaction Executed and Committed with Error CONSTRAINT_NOT_SATISFIED\nExecution failed with status: metadata and code bundle mismatch: unregistered dependency: '5ba9fa393e4840013932ecceb243ae2e6c38a558444fe777ebe8e0550c4c4e18::aptos_horses_publisher_signer'"
}
```

You can simply fix this through a fairly standard flow, where you provide a placeholder name.

```
[addresses]
publisher = "_" # This is a placeholder, and has to be set at runtime

[dev-addresses]
publisher = "0x5ba9fa393e4840013932ecceb243ae2e6c38a558444fe777ebe8e0550c4c4e18"
```

Then you can publish it by setting the name e.g.

```bash
aptos move publish --named-addresses publisher=default
```

Or by resource account:
```bash
aptos move create-resource-account-and-publish-package --address-name publisher --seed whatever
```

This particular contract, then fails due to an error with resource account, but this should unblock you here.
